### ISSUE TITLE
[5] Design system for backups

### ISSUE THREAD

**Issue by jhammock** [EOL/tramea#134]

**Comment by jhammock** [comment]
placeholder for @JRice

**Comment by JRice** [comment]
The main focuses for backups are:
1. Restoring the host machine, including the rebuild (OS and main tools) and the bootstrap (**chef** and **docker**)
2. Restoring data, which can mean either **MySQL**, **Solr**, or **Virtuoso** (memcached and redis are transient).
3. Restoring files, which is mostly concerned with **resources** and **media**, but less critically, **logs**, and the clusterflurk that is the **harvesting machine**, which has a myriad "special" places to hide files (in fact, these are not well understood or enumerated).

**Chef**
This is pretty well backed up via github, but restoring it would be a pain. For now, this is acceptable.

**Docker**
This is entirely reliant on dockerhub, which carries its own risks. For now, though, this is acceptable.

**MySQL**
As long as we have functioning slaves, we can continue to use one of them to perform "live" backups. We 'll need a cron job to run a particular script to run the backups. It needs to stop the slave before starting and it must ALWAYS restart the slave after it exits (whether or not there was an error). We will want logs for when backups occur and notifications for when they fail.

**Solr**
I have no idea how to do this. The main problem is that it should really be in sync with the database backups, lest we have to rebuild _everything_, which takes an inordinately long time (days at least). This requires serious thought. _After some thought:_ I actually don't think this is _super_ important, with one glaring exception: publishing resources. I suggest we backup Solr after every "harvest batch", and leave it at that—it will be _usable,_ to an acceptable extent. It might not be perfect (manual curation changes might not be reflected), but the exceptions should be pretty rare, and a "reindex" command would fix those cases. For an emergency restore, this should be acceptable.

**Virtuoso**
This seems to be quite simple: copy the files. It's best to do this when the database isn't being written to, but we really only write to it during harvests. Perhaps if we make this part of the "harvesting" queue (or something like that), we should be fine.

**Resources**
Stored in /opt/resources on the harvesting machine. Not at all backed up, at present. These are "just files," so backups will be easy. Fine to just tar the whole directory.

**Media**
Woof. This is the largest piece; I think we're at 1.5 TB, if I recall. _[update: No, it's more like *_12TB*_, sigh. We will have to back this up externally; we do not have the space.]_ I'm not sure what's "best" for backing this stuff up. Some of it is effectively stale (we never remove old icons, failed uploads, or unpublished media), a lot of it is rarely indexed (old versions of media), and the rest, ridiculously, is split between two places: up to a certain date, they live on eol1 (`/eol/data/content/`), and after that date, they live on the harvesting machine(`/eol/data/var/www/content/`). I really, really hate that.

**Logs**
These live in the /eol/data/*\* directories on each machine (different subdir for each container on the host), and are more or less important based on which machine they're on. It's not critical that these get backed up, but it would be nice to have. For now, I think it's acceptable to ignore them, though. Once we have a regular sysadmin, though, I'd like to add these to the mix.

**Harvesting Machine**
Sigh. It's easiest to say that this is "all the files on the machine that aren't part of the docker container or the git repository or the resources or the (recent) media." And that being the _simple_ version should give you an idea of how ugly this is. :( I am not sure what to do about this. I'm not sure how important which files are, other than to say we've already had a few bugs reported for things that failed since April because of this. It's a hair situation, and we can't ignore it, but this should be lower-priority. It's the 80% of the work for the 20% of the result. :(

**Comment by JRice** [comment]
For posterity, here are the git repos that I have on slave2 (the chef server):

```
dotfiles           eol-base-private-cookbook      eol-logrotate-wrapper-cookbook  eol-users-cookbook  mysql   varnish
eol                eol-docker-cookbook            eol-sensu-wrapper-cookbook      harvest             resque
eol-base-cookbook  eol-iptables-wrapper-cookbook  eol-si-repo                     memcached           solr
```

And several of those are symlinked TO the cookbooks directory for eol-si-repo:

```
lrwxrwxrwx  1 jrice jrice   34 Apr 11  2015 eol-base -> /home/jrice/git/eol-base-cookbook/
lrwxrwxrwx  1 jrice jrice   42 Apr  9  2015 eol-base-private -> /home/jrice/git/eol-base-private-cookbook/
lrwxrwxrwx  1 jrice jrice   36 Apr 11  2015 eol-docker -> /home/jrice/git/eol-docker-cookbook/
lrwxrwxrwx  1 jrice jrice   46 Apr  8  2015 eol-iptables-wrapper -> /home/jrice/git/eol-iptables-wrapper-cookbook/
lrwxrwxrwx  1 jrice jrice   46 Jul  6  2015 eol-logrotate-wrapper -> /home/jrice/git/eol-logrotate-wrapper-cookbook
lrwxrwxrwx  1 jrice jrice   43 Apr 12  2015 eol-sensu-wrapper -> /home/jrice/git/eol-sensu-wrapper-cookbook/
```

**Comment by JRice** [comment]
Also for posterity, it's worth reading the "Smithsonian (SI) Servers - What We Know" in the Google Docs folder for "DevOps." I don't want to link to it here, even though it's secured... it's got a lot of juicy information in it (no passwords, but lots of IPs and names and things). :\

**Comment by JRice** [comment]
Storing the backed-up data is another consideration.

We can get really fancy here, depending on the requirements. 

Daily backups would be great, and I think we should aim for that. The worst, time-wise, is the SQL dump, and that takes a few hours, so that should be acceptable, I think.

I'd suggest we keep the latest five backups, regardless of when they were (in case one is corrupt); the 1st backup of the month; and the 1st backup of the year. We'll need a script to do the logic.

I suggest we label backups based on the names I've used here, and just keep them all in once place, for simplicity, so it might look something like this:

```
mysql-2016-01-01.gz
mysql-2016-02-01.gz
mysql-2016-02-24.gz
mysql-2016-02-19.gz
mysql-2016-02-20.gz
mysql-2016-02-21.gz
mysql-2016-02-22.gz
mysql-2016-02-23.gz
mysql-2016-02-24.gz
solr-2016-01-01.gz
[etc...]
virtuoso-2016-01-01.gz
[etc, etc...]
```

**Comment by JRice** [comment]
Storing the media as a complete gzip every day is **not** a good idea (and without using tape, not possible). We'll just want that to be a cumulative backup, probably by year for complete years, and by month for the most recent year's previous months, and by day for the most recent month. That's a bit of a pain, but entirely scriptable. So it would look like this (and I suggest we put these in their own subdir called 'media,' despite the redundancy):

```
[older years here...]
media-2014.tgz
media-2015.tgz
media-2016-01-01.tgz
media-2016-02-01.tgz
media-2016-02-02.tgz
media-2016-02-03.tgz
media-2016-02-04.tgz
[etc, up to today]
```

**Comment by JRice** [comment]
I'm not claiming this is clean code, but here's a (non-removing) version of the file to clean up according to those specs (actually not quite—it keeps an additional file in some border cases, but I don't care):

``` ruby
contents = Dir.entries(".").grep(/^\w+-\d\d\d\d-\d\d-\d\d.t?gz$/).sort

# puts contents.map(&:to_s).join("\n")

parts = contents.map { |c| (t, y, m, d) = c.split("-") ; {file: c, type: t, year: y, month: m, day: d} }

types = parts.group_by { |p| p[:type] }

types.each do |type, years|
  puts "TYPE: #{type}:"
  years.group_by { |p| p[:year] }.each do |year, months|
    if months.size > 5
      last_five = months.slice!(-5..-1)
      puts "  Keeping the last five backups, always:"
      last_five.each do |file|
        puts "    #{file[:file]}"
      end
    else
      puts "  There are five or less files in #{year}, keeping all:"
      months.each do |file|
        puts "    #{file[:file]}"
      end
      next
    end
    if months.empty?
      puts "  Nothing left."
      next
    end
    puts "  Keeping first file of the year:\n    #{months.first[:file]}"
    months.shift
    if months.empty?
      puts "  ...That's the only #{type} file in #{year}."
      next
    end
    months.group_by { |p| p[:month] }.each do |month, days|
      first = days.shift
      puts "  Keeping first of month:\n    #{first[:file]}"
      if days.empty?
        puts "  ...That's the only #{type} file in the month #{month}."
        next
      end
      days.each do |day|
        puts "  ** Can remove: #{day[:file]}"
      end
    end
  end
end
```

**Comment by JRice** [comment]
For posterity, here is a bash script for backing up the database:

```
#!/bin/bash

trap 'mysqladmin -h eol-db-slave2 start-slave' EXIT

export FORMATTED_DATE=`date +%Y-%m-%d`
export DUMPFILE="mysql-$FORMATTED_DATE.gz"
mysqladmin -h eol-db-slave2 stop-slave
mysql -e "show slave status\G" | grep Master_Log
export FILE_NAME=`mysql -e "show slave status\G" | grep Relay_Master_Log_File | awk '{print $2;}'`
export FILE_POS=`mysql -e "show slave status\G" | grep Exec_Master_Log_Pos | awk '{print $2;}'`
echo "SET SESSION SQL_LOG_BIN=0;" | gzip -1 > $DUMPFILE
if mysqldump -h eol-db-slave2 --opt --single-transaction --no-autocommit --skip-lock-tables --flush-logs --master-data=2 \
  --all-databases | gzip -1 >> $DUMPFILE ; then
  # It worked
  echo "On new slave: CHANGE MASTER TO MASTER_LOG_FILE = '$FILE_NAME', MASTER_LOG_POS = $FILE_POS;"
  scp $DUMPFILE eol-archive:~/archive/.
else
  # It failed
  exit 1
fi
# NOTE this fires the "trap" at the top of the file.
exit 0

# NOTE: Recommended command to watch A RESTORE of the database:
# watch -n 60 "bash -c 'ps -ef | grep unzip' && mysql -A -e 'show tables;' eol_production | tail -n 3"
```

**Comment by JRice** [comment]
Solr backups remain a mystery to me. AFAICT, you need to stop the server to get a real backup. **We can't do that.**

It's possible there are other solutions, but there is no documentation for it ...on the web. We're using an ancient version of it.

This book might have a solution for our version: https://www.packtpub.com/big-data-and-business-intelligence/apache-solr-31-cookbook ...is that something folks at the BA have access to?

**Comment by JRice** [comment]
I added a "backup.sh" script on the virtuoso server, but only as the root user... and it moves the file into the ~jrice dir, which is suboptimal. Really, I'd like to be able to backup as jrice (or a new "backup" user, if I'm feeling productive)... but ... later. Not 100% sure what the best way to give access to a backup user is. sudo didn't work for `gtar`, for some reason. I suppose I could run `sudo ~root/backup.sh` and then handle the resulting file.

I also tried setting up ssh on this machine to be able to get directly to eol-archive. It was as I thought: `ssh-keygen`, copy the pubkey over to archive's `authorized_keys`, and it works fine.

So we'll want to do that on other key machines, too.

**Comment by JRice** [comment]
Okay, I am "closing" this ticket since I've done enough preliminary work to iron out the details in another iteration. Sigh.

**Comment by JRice** [comment]
One last observation: before deciding the media backups would be too large, I tar-gzipped the 2014 directory. ...I decided to let that process finish, to see how large it would end up. It was 1.5 TB. (I then removed it because I have nowhere to put it!)

So we're going to need a very large place to put all of the yearly directories (they won't all be that size, I chose 2014 because I assume it would be the largest—since we weren't really harvesting for most of 2015).
